sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XXE/XmlReaderSettingsValidator.cs

```diff
@@ -0,0 +1,106 @@
+
+            var assignments = invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>()
+                .DescendantNodes()
+                .OfType<AssignmentExpressionSyntax>()
+                .Where(assignment => assignment.Left is MemberAccessExpressionSyntax memberAccess &&
+                                     IsXmlReaderSettings(memberAccess.Expression, semanticModel) &&
+                                     settings.Equals(semanticModel.GetSymbolInfo(memberAccess.Expression).Symbol));
+
+            foreach (var assignment in objectCreationAssignments.Union(assignments))
+            {
+                var name = assignment.Left.GetName();
+
+                if (name =="ProhibitDtd" || name == "DtdProcessing")
+                {
+                    unsafeDtdProcessing = IsXmlResolverDtdProcessingUnsafe(assignment, semanticModel);
+                }
+                else if (assignment.Left.NameIs("XmlResolver"))
+                {
+                    unsafeResolver = IsXmlResolverAssignmentUnsafe(assignment, semanticModel);
+                }
+            }
+
+            return unsafeDtdProcessing && unsafeResolver;
+        }
+
+        private static bool IsXmlReaderSettings(ExpressionSyntax expressionSyntax, SemanticModel semanticModel) =>
+            semanticModel.GetTypeInfo(expressionSyntax).Type.Is(KnownType.System_Xml_XmlReaderSettings);
+
+        private static ObjectCreationExpressionSyntax GetObjectCreation(ISymbol symbol, InvocationExpressionSyntax invocation, SemanticModel semanticModel) =>
+            symbol.Locations
+                .SelectMany(location => GetDescendantNodes(location, invocation).OfType<ObjectCreationExpressionSyntax>())
+                .FirstOrDefault(objectCreation => objectCreation.Initializer != null && IsXmlReaderSettings(objectCreation, semanticModel));
+
+        private static IEnumerable<SyntaxNode> GetDescendantNodes(Location location, InvocationExpressionSyntax invocation) =>
+            location.SourceTree?.GetRoot()?.FindNode(location.SourceSpan).DescendantNodes() ??
+            invocation.FirstAncestorOrSelf<MethodDeclarationSyntax>()?.DescendantNodes() ??
+            Enumerable.Empty<SyntaxNode>();
+
+        private static bool IsXmlResolverDtdProcessingUnsafe(AssignmentExpressionSyntax assignment, SemanticModel semanticModel) =>
+            semanticModel.GetConstantValue(assignment.Right).Value switch
+            {
+                false => true,
+                (int)DtdProcessing.Parse => true,
+                _ => false
+            };
+
+        private static bool IsXmlResolverAssignmentUnsafe(AssignmentExpressionSyntax assignment, SemanticModel semanticModel)
+        {
+            if (assignment.Right.IsKind(SyntaxKind.NullLiteralExpression))
+            {
+                return false;
+            }
+
+            var symbol = semanticModel.GetSymbolInfo(assignment.Right).Symbol;
+            var type = symbol.GetSymbolType() ?? symbol.ContainingType;
```

Comment:
I don't understand this line - is it the type of the symbol or the type of the containing type?

Reply:
Indeed this is not straight forward. If XmlResolver is set during constructor initialization, `GetSymbolType()` will return null so we have to rely on ContainingType to get the type.

I will add a comment.